pkg/icingadb/schema.go

```diff
@@ -15,27 +17,74 @@ const (
 	expectedPostgresSchemaVersion = 4
 )
 
-// CheckSchema asserts the database schema of the expected version being present.
-func CheckSchema(ctx context.Context, db *database.DB) error {
-	var expectedDbSchemaVersion uint16
+var (
+	// ErrSchemaNotExists implies that no Icinga DB schema has been imported.
+	ErrSchemaNotExists = fmt.Errorf("no database schema exists")
+
+	// ErrSchemaMismatch implies an unexpected schema version, most likely after Icinga DB was updated but the database
+	// missed the schema upgrade.
+	ErrSchemaMismatch = fmt.Errorf("unexpected database schema version")
+)
+
+// CheckSchema verifies the correct database schema is present.
+//
+// This function returns the following error types, possibly wrapped:
+//   - If no schema exists, the error returned is ErrSchemaNotExists.
+//   - If the schema version does not match the expected version, the error returned is ErrSchemaMismatch.
+//   - Otherwise, the original error is returned, e.g., for general database problems.
+func CheckSchema(ctx context.Context, db *database.DB, databaseName string) error {
+	var (
+		tableSchema             string
+		expectedDbSchemaVersion uint16
+	)
 	switch db.DriverName() {
 	case database.MySQL:
+		tableSchema = databaseName
 		expectedDbSchemaVersion = expectedMysqlSchemaVersion
+
 	case database.PostgreSQL:
+		tableSchema = "public"
 		expectedDbSchemaVersion = expectedPostgresSchemaVersion
-	}
 
-	var version uint16
+	default:
+		return fmt.Errorf("unsupported database driver %q", db.DriverName())
+	}
 
+	var tableSchemaCount int
 	err := retry.WithBackoff(
 		ctx,
-		func(ctx context.Context) (err error) {
+		func(ctx context.Context) error {
+			// The following query should return either 0 or 1 depending on the existence of the icingadb_schema table.
+			// Unfortunately, a "SELECT 1" query does not work, because at least the PostgreSQL driver does not raise an
+			// error for empty results.
+			query := db.Rebind("SELECT COUNT(*) FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA=? AND TABLE_NAME='icingadb_schema'")
+			if err := db.QueryRowxContext(ctx, query, tableSchema).Scan(&tableSchemaCount); err != nil {
+				return database.CantPerformQuery(err, query)
+			}
+			return nil
+		},
+		retry.Retryable,
+		backoff.NewExponentialWithJitter(128*time.Millisecond, 1*time.Minute),
+		db.GetDefaultRetrySettings())
+	if err != nil {
+		return errors.Wrap(err, "can't verify existence of database schema table")
+	}
+	if tableSchemaCount == 0 {
+		return ErrSchemaNotExists
+	}
+	if tableSchemaCount != 1 {
+		return fmt.Errorf("no or one 'icingadb_schema' tables are expected, found %d", tableSchemaCount)
+	}
```

Comment:
I am still not happy with `tableSchemaCount`. There can't be anything other than `0` or `1` here. We have the same functionality in our Kubernetes daemon:

https://github.com/Icinga/icinga-kubernetes/blob/main/cmd/icinga-kubernetes/main.go#L590

I suggest moving this to `DB#HasTable()`, with the adjustment to handle both PostgreSQL and MySQL.